network/src/protocols/stream/mod.rs

```diff
@@ -0,0 +1,226 @@
+            NetworkMessage::Error(_) => panic!("StreamHeader with Error should be rejected"),
+            NetworkMessage::RpcRequest(request) => request.raw_request.append(raw_data),
+            NetworkMessage::RpcResponse(response) => response.raw_response.append(raw_data),
+            NetworkMessage::DirectSendMsg(message) => message.raw_msg.append(raw_data),
+        }
+        Ok(self.current_fragment_id == self.num_fragments)
+    }
+}
+
+pub struct OutboundStream {
+    request_id_gen: U32IdGenerator,
+    max_frame_size: usize,
+    max_message_size: usize,
+    stream_tx: Sender<MultiplexMessage>,
+}
+
+impl OutboundStream {
+    pub fn new(
+        max_frame_size: usize,
+        max_message_size: usize,
+        stream_tx: Sender<MultiplexMessage>,
+    ) -> Self {
+        // some buffer for headers
+        let max_frame_size = max_frame_size - 64;
+        assert!(
+            max_frame_size * u8::MAX as usize >= max_message_size,
+            "Stream only supports maximum 255 chunks, frame size {}, message size {}",
+            max_frame_size,
+            max_message_size
+        );
+        Self {
+            request_id_gen: U32IdGenerator::new(),
+            max_frame_size,
+            max_message_size,
+            stream_tx,
+        }
+    }
+
+    pub fn should_stream(&self, message: &NetworkMessage) -> bool {
+        message.data_len() > self.max_frame_size
+    }
+
+    pub async fn stream_message(&mut self, mut message: NetworkMessage) -> anyhow::Result<()> {
+        ensure!(
+            message.data_len() <= self.max_message_size,
+            "Message length {} exceed size limit {}",
+            message.data_len(),
+            self.max_message_size,
+        );
+        ensure!(
+            message.data_len() >= self.max_frame_size,
+            "Message length {} is smaller than frame size {}, should not go through stream",
+            message.data_len(),
+            self.max_frame_size,
+        );
+        let request_id = self.request_id_gen.next();
+        let rest = match &mut message {
+            NetworkMessage::Error(_) => {
+                unreachable!("NetworkMessage::Error should always fit in a single frame")
+            }
+            NetworkMessage::RpcRequest(request) => {
+                request.raw_request.split_off(self.max_frame_size)
+            }
+            NetworkMessage::RpcResponse(response) => {
+                response.raw_response.split_off(self.max_frame_size)
+            }
+            NetworkMessage::DirectSendMsg(message) => {
+                message.raw_msg.split_off(self.max_frame_size)
+            }
+        };
+        let chunks = rest.chunks(self.max_frame_size);
+        ensure!(
+            chunks.len() <= u8::MAX as usize,
+            "Number of fragments overflowed"
+        );
+        let header = StreamMessage::Header(StreamHeader {
+            request_id,
+            num_fragments: chunks.len() as u8,
+            message,
```

Comment:
Can you explain why the message is part of the header?

Reply:
it both carries the first fragment and metadata like the protocol_id etc